### PR TITLE
NO MORE HALF MEASURES, CLUSTER

### LINKED ISSUE #10694: Pointing at things makes the pointer arrow show up under the object
(WEB REPORT BY: justsumguy REMOTE: 172.93.109.202:7777)

> Revision (Should be above if you're viewing this from ingame!)

1fe6a823d2f983f71a81a0c2d80d1ebac15107fe 

> General description of the issue

Point at something bigger than the pointer icon and it 

> What you expected to happen

To see what is being pointed at

> What actually happened

Pointer is hidden behind the object 

> Steps to reproduce if possible

Point at stuff(large machines especially)

### LINKED ISSUE #10756: Air pipes layered below disposal pipes
(WEB REPORT BY: sprok REMOTE: 172.93.109.202:7777)

> Revision (Should be above if you're viewing this from ingame!)
> c4cd8f15fa9750fe4f974a75e61e8383a0b8d681 
> 
> General description of the issue
> Atmos pipes are layered below disposal pipes
> 
> What you expected to happen
> Air pipes to be layered above disposal pipes
> 
> What actually happened
> Need to alt-click to access the air pipes
> 
> Steps to reproduce if possible
> Pull up floor tiles somewhere that there are air pipes and disposal pipes on the same tile

### LINKED ISSUE #10804: item layering is borked
(WEB REPORT BY: cumoverme REMOTE: 172.93.109.202:7777)

> Revision
> 9709e5d58d69e0e097496e3af7870ca97eb44066
> 
> General description of the issue
> https://u.pomf.is/oovirc.png
> Items you hold are being shown behind everything else, very perceptible with eswords and corgis.

### PR DESCRIPTION
**Pull request by ghost**
this is my largest PR to date in terms of files changed, and sum of lines changed.
why did I agree to do planes

ADDS: 
- layers are now defines (this is a big deal)
- added some documentation to _planes+layers to help people I guess

TESTING THE FIX FOR:
(not having an X doesn't mean I haven't fixed it, just means I haven't yet tested without a doubt whether or not it IS fixed)
- [X] DISPOSAL PIPES VERSUS AIR PIPES - fixes #10756
- [X] CABLES VERSUS TERMINAL 
- [X] PLAYER VERSUS ITEMS IN HAND - fixes #10804
- [x] POINTER VERSUS OTHER THINGS - fixes #10694
- [x] FORCEFIELDS INVISIBLE (you guys did realise this isn't a planes issue right)

NEW BUGS FOUND:

(if box is checked, the bug has been fixed.) 
- [x] /obj/screen/inventory used the wrong layer, layer = 19.
- [x] all hud objects appear less responsive  (probably to do with using my laptop trackpad, actually.)
- [x] if you slip on the clown's PDA, you will appear underneath it, rather than on top. - will need to re-evaluate the object plane. (never mind, I just gave blankets their own plane and bumped up lying humans. should work fine)
- [x] damage overlays are above the HUD, rather than below. Should give them their own plane, I think. (done)
- [x] cables are on the wrong plane.
- [x] pipes are on the wrong plane. 
- [x] Privacy shutters are on the wrong plane.
- [x] Forgot to put humans on the lying human plane when they are lying.
- [x] Forgot to put humans on the human plane when they are not.
- [x] Status Displays are on the wrong plane
- [x] MoMMIs appear blind when hiding
- [x] Door layering is wacky.
- [x] MoMMIs cannot unhide.
- [x] density of poddoors New() wrong way around
- [x] side windows appeared above poddoors
- [x] UNFIXABLE: North facing chairs appear above objects that they shouldn't. (This is unfixable. - You cannot have them appear above humans and below objects which appear below humans. - I imagine that this was already a bug.)
- [x] Cryocells were on the wrong layer.
- [x] Catwalks were on the wrong plane.
- [x] Showers & Curtains behaved oddly when combined. Fixed.
- [x] signs did not appear above windows. They do now.
- [x] Monkies did not layer correctly when lying down.
- [x] Accidentally pushed the wrong map.
- [x] Wrong window was on top of the other.
- [x] Alien nests layered incorrectly.
- [x] Molten items did not layer correctly.

100% TESTED:
- [X] SECURITY CAMERAS AND LIGHTBULBS APPEAR ABOVE HUMANS
- [X] NOTHING APPEARS ABSOLUTELY FUBAR SO FAR
- [X] Comfy Chairs, Secways, Janicarts & Chairs appear normal
- [X] Turrets seem normal
- [X] Labcoats work
- [X] Bags work
- [X] Pockets work
- [X] MoMMIs hats still work, their eyes still appear above darkness, and dropping from them layers correctly.
- [X] The Holomap renders appropriately. This should be turned into a plane.
- [X] Mesons still work
- [X] In-hand cyborgs work as intended.
- [X] oh wow the engineering borg sprite has spooky eyes
- [X] People can still hide behind plants
- [X] Blobs seem to still layer correctly.
- [x] corgi and carp hearts layer correctly
- [x] cages layer correctly
- [x] Sorry, JustSumGuy, this definitely does not fix cameras.
- [x] cobwebs layer correctly.
- [x] liquid fuel layers correctly

TODO:
- [x] will need to rethink the OBJ_PLANE, it's too big and all encompassing, see the clown-PDA bug (never mind, I just gave blankets their own plane and bumped up lying humans. should work fine)
- [x] Expand the tested list...?
- [x] Doors handled layers really weirdly, DOOR_MOD and such. Make it use var/open_layer & var/closed_layer instead.
- [x] Read through my changes and try and catch out mistakes, such as putting OBJ_PLANE where I meant OBJ_LAYER. I'm almost certain I've done this somewhere.
- [X] I don't trust turrets, they had really funky layers when I was looking at them, let's take a look if I did it correctly. (Seem okay.)
- [x] Look at every single department on Box station and check if anything is amiss.
- [x] Do the same with Defficiency.
- [x] fix the code issues I pointed out

This was all done by hand, I don't know fancy replace code

### CODE REVIEW

**Comment by Intigracy** [comment]
This branch has conflicts that must be resolved

**Comment by Intigracy** [comment]
> layers are now defines FIXES: I don't know, it might fix things and it might break things, I haven't tested it.
> I haven't tested it
> haven't tested

WHO WOULD HAVE GUESSED

**Comment by 9600bauds** [comment]
https://www.youtube.com/watch?v=vNadq3lyD7M

File: code/ATMOSPHERICS/components/unary/unary_base.dm [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /obj/machinery/atmospherics/unary
 	dir = SOUTH
 	initialize_directions = SOUTH
-	layer = 2.45 // Cable says we're at 2.45, so we're at 2.45.  (old: TURF_LAYER+0.1)
+	layer = UNARY_PIPE_LAYER
```
Comment (Intigracy) [inline code/ATMOSPHERICS/components/unary/unary_base.dm]: why do you need so many planes for specific things?

this is why planes AND layers exist

**Comment by ScrambledSteak** [comment]
> celt didn't test a giant fucking PR

GEE WHO WOULD HAVE GUESSED
_revert hypothermia when_

**Comment by PJB3005** [comment]
:airplane:

**Comment by unid15** [comment]
:airplane: :office: :office:

**Comment by PJB3005** [comment]
:bomb: :boom: :fire: :fire_engine:

**Comment by ghost** [comment]
now that it's compiled and rebased, I'll do a week's robust testing on it.
please do not merge any PRs that mess with layers/planes in that time. 
something something merge priority.

**Comment by ghost** [comment]
I'm done with this for the night. Please merge #11196 so I can test more offline, and uh, remove the conflicts tag from this.

**Comment by ghost** [comment]
Does not fix #10901 - I know how to fix this, combined with heavily optimise it, pointed out to me by @clusterfack and it'd be unatomic to have that fix in this one.

**Comment by ghost** [comment]
Tell me what to test or forever hold your peace. 
I'll read through my code line by line and see if I can find any obvious errors.

**Comment by ghost** [comment]
my prediction is ready to merge by tuesday, when I get wifi.

**Comment by JustSumBody** [comment]
Out of morbid curiosity, does this somehow fix nude cameras?

**Comment by ghost** [comment]
@JustSumBody Don't believe so. That's not a planes issue as far as I am aware.

**Comment by ghost** [comment]
I'll read through all of my changes tomorrow, and this will be ready to merge by Tuesday.

File: code/datums/diseases/wizarditis.dm [hunk 2]
```diff
@@ -62,24 +62,21 @@ STI KALY - blind
 				if(H.head)
 					H.drop_from_inventory(H.head)
 				H.head = new /obj/item/clothing/head/wizard(H)
-				H.head.layer = 20
-				H.shoes.plane = PLANE_HUD
```
Comment (ghost) [inline code/datums/diseases/wizarditis.dm:66]: ehehehehehehehehhhhh... It's not what it looks like!

File: code/game/objects/effects/decals/Cleanable/fuel.dm [hunk 3]
```diff
@@ -2,7 +2,6 @@
 	//Liquid fuel is used for things that used to rely on volatile fuels or plasma being contained to a couple tiles.
 	icon = 'icons/effects/effects.dmi'
 	icon_state = "fuel"
-	layer = TURF_LAYER+0.2
```
Comment (ghost) [inline code/game/objects/effects/decals/Cleanable/fuel.dm:5]: Todo: Test this.

File: code/game/objects/effects/decals/Cleanable/misc.dm [hunk 4]
```diff
@@ -57,7 +53,7 @@
 	desc = "Somebody should remove that."
 	density = 0
 	anchored = 1
-	layer = 3
+	layer = OBJ_LAYER
```
Comment (ghost) [inline code/game/objects/effects/decals/Cleanable/misc.dm]: Test this.

File: code/game/objects/effects/decals/Cleanable/misc.dm [hunk 5]
```diff
@@ -66,7 +62,7 @@
 	desc = "It looks like a melted... something."
 	density = 0
 	anchored = 1
-	layer = 3
+	layer = OBJ_LAYER
```
Comment (ghost) [inline code/game/objects/effects/decals/Cleanable/misc.dm:64]: Ditto

File: code/game/objects/effects/decals/Cleanable/misc.dm [hunk 6]
```diff
@@ -75,7 +71,7 @@
 	desc = "Somebody should remove that."
 	density = 0
 	anchored = 1
-	layer = 3
+	layer = OBJ_LAYER
```
Comment (ghost) [inline code/game/objects/effects/decals/Cleanable/misc.dm]: Ditto.

File: code/game/objects/structures/cage.dm [hunk 7]
```diff
@@ -51,11 +51,11 @@
 
 	if(cover_state == C_CLOSED)
 		var/image/cover_overlay = image('icons/obj/cage.dmi', icon_state = "cage_cover", layer = OBJ_LAYER)
-		cover_overlay.plane = PLANE_OBJ
+		cover_overlay.plane = OBJ_PLANE
 		overlays += cover_overlay
 	else if(door_state == C_CLOSED) //Door is only visible when the cover is open
 		var/image/door_overlay = image('icons/obj/cage.dmi', icon_state = "cage_door", layer = MOB_LAYER + 0.5) //Above mobs
-		door_overlay.plane = PLANE_MOB
+		door_overlay.plane = MOB_PLANE
```
Comment (ghost) [inline code/game/objects/structures/cage.dm]: Test this, change to above human plane.

File: code/game/objects/structures/electricchair.dm [hunk 8]
```diff
@@ -10,7 +10,7 @@
 /obj/structure/bed/chair/e_chair/New()
 	..()
 	var/image/over = image('icons/obj/objects.dmi', src, "echair_over", MOB_LAYER + 1, dir)
-	over.plane = PLANE_MOB
+	over.plane = MOB_PLANE
```
Comment (ghost) [inline code/game/objects/structures/electricchair.dm]: Above human plane

File: code/game/objects/structures/electricchair.dm [hunk 9]
```diff
@@ -46,7 +46,7 @@
 /obj/structure/bed/chair/e_chair/rotate()
 	..()
 	overlays.len = 0
-	overlays += image('icons/obj/objects.dmi', src, "echair_over", MOB_LAYER + 1, dir,plane = PLANE_MOB)	//there's probably a better way of handling this, but eh. -Pete
+	overlays += image('icons/obj/objects.dmi', src, "echair_over", MOB_LAYER + 1, dir,plane = MOB_PLANE)	//there's probably a better way of handling this, but eh. -Pete
```
Comment (ghost) [inline code/game/objects/structures/electricchair.dm]: Above human plane

File: code/game/objects/structures/fullwindow.dm [hunk 10]
```diff
@@ -10,7 +10,7 @@
 	icon_state = "window0" //Specifically for the map
 	sheetamount = 2
 	mouse_opacity = 2 // Complete opacity //What in the name of everything is this variable ?
-	layer = 3.21 // Windows are at 3.2.
+	layer = FULL_WINDOW_LAYER
```
Comment (ghost) [inline code/game/objects/structures/fullwindow.dm:13]: Was this supposed to be above or below side windows? Test.

File: code/game/objects/structures/stool_bed_chair_nest/alien_nests.dm [hunk 11]
```diff
@@ -62,7 +62,7 @@
 	lock_atom(M, /datum/locking_category/bed/nest)
 	src.add_fingerprint(user)
 	var/image/nest_covering = image(icon,"nest-covering",MOB_LAYER)
-	nest_covering.plane = PLANE_MOB
+	nest_covering.plane = MOB_PLANE
```
Comment (ghost) [inline code/game/objects/structures/stool_bed_chair_nest/alien_nests.dm]: Uh, test.

File: code/game/turfs/simulated/floor_types.dm [hunk 12]
```diff
@@ -262,7 +262,7 @@
 /turf/simulated/floor/beach/water/New()
 	..()
 	var/image/water = image("icon"='icons/misc/beach.dmi',"icon_state"="water5","layer"=MOB_LAYER+0.1)
-	water.plane = PLANE_MOB
+	water.plane = MOB_PLANE
```
Comment (ghost) [inline code/game/turfs/simulated/floor_types.dm]: Change.

File: code/game/turfs/simulated/floor.dm [hunk 13]
```diff
@@ -646,7 +646,7 @@ turf/simulated/floor/update_icon()
 	if((icon_state != "cult")&&(icon_state != "cult-narsie"))
 		name = "engraved floor"
 		icon_state = "cult"
-		turf_animation('icons/effects/effects.dmi',"cultfloor",0,0,MOB_LAYER-1,anim_plane = PLANE_OBJ)
+		turf_animation('icons/effects/effects.dmi',"cultfloor",0,0,MOB_LAYER-1,anim_plane = OBJ_PLANE)
 	return
```
Comment (ghost) [inline code/game/turfs/simulated/floor.dm:650]: Trailing return, remove.

File: code/modules/mob/living/carbon/monkey/monkey.dm [hunk 14]
```diff
@@ -613,8 +613,6 @@
 
 /mob/living/carbon/monkey/reset_layer()
 	if(lying)
-		plane = PLANE_OBJ
-		layer = MOB_LAYER - 0.1 //so we move under bedsheets
+		plane = HIDING_MOB_PLANE
```
Comment (ghost) [inline code/modules/mob/living/carbon/monkey/monkey.dm]: Gonna need to give monkies a lying mob plane - this is inappropriately placed

File: code/modules/mob/living/simple_animal/hostile/carp.dm [hunk 15]
```diff
@@ -150,7 +150,7 @@
 					friends.Add(user)
 					to_chat(user, "<span class='info'>You have gained \the [src]'s trust.</span>")
 					var/image/heart = image('icons/mob/animal.dmi',src,"heart-ani2",MOB_LAYER+1)
```
Comment (ghost) [inline code/modules/mob/living/simple_animal/hostile/carp.dm]: remove that layer

File: vgstation13.dme [hunk 16]
```diff
@@ -2005,7 +2005,8 @@
 #include "interface\web\interface.dms"
 #include "maprendering\maprendering.dm"
 #include "maps\_map.dm"
-#include "maps\tgstation.dm"
+#include "maps\test_tiny.dm"
```
Comment (ghost) [inline vgstation13.dme]: oh dear oh dear oh dear

**Comment by ghost** [comment]
once that is done I can say with certainty that planes is ready to merge.

**Comment by Arthurdentist** [comment]
So does this mean we can start working on Train Station next?

Once we get the Steam Engine PR done everything will be golden.

**Comment by ghost** [comment]
Ready to merge.

**Comment by ghost** [comment]
merge valids

**Comment by clusterfack** [comment]
@JustSumBody Has co-signed this PR in the moderately likely chance that @DrCelt mysteriously vanishes, again.

**Pull request merged**